Docs: add "how to share data" (#4458)
aiidateam/aiida-core: docs/source/howto/share_data.rst

    If an outdated archive version is detected during import, you will be prompted to confirm automatic migration of the archive file.

    You can also use ``verdi export migrate`` to update existing archive files once and for all.

.. tip:: In order to get a quick overview of an archive file *without* importing it into your AiiDA profile, use ``verdi export inspect``:

    .. code-block:: console

        $ verdi export inspect sssp-efficiency.aiida
        --------------  -----
        Version aiida   1.2.1
        Version format  0.9
        Computers       0
        Groups          0
        Links           0
        Nodes           85
        Users           1
        --------------  -----

    Note: For archive versions 0.2 and below, the overview may be inaccurate.

.. _how-to:data:serve:

Serving data through the REST API
=================================

The AiiDA REST API allows to query your AiiDA database over HTTP(S).

.. note::

    As of October 2020, the AiiDA REST API does *not* yet support workflow management.
    This feature is, however, part of the `AiiDA roadmap <https://github.com/aiidateam/aiida-core/wiki/AiiDA-release-roadmap>`_.

> 

Reviewer: Please add a link here:
```
.. _how-to:data:serve:launch:
```
I'd like to reference it

Author: thanks, will do!

I just fixed the section name, it will be `how-to:share:serve:launch`

Launching the REST API
^^^^^^^^^^^^^^^^^^^^^^

Start serving data from your default AiiDA profile via the REST API:

.. code-block:: console

    $ verdi restapi
     * REST API running on http://127.0.0.1:5000/api/v4
     * Serving Flask app "aiida.restapi.run_api" (lazy loading)
     * Environment: production
       WARNING: This is a development server. Do not use it in a production deployment.
       Use a production WSGI server instead.
     * Debug mode: off
     * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)

The REST API is now running on port ``5000`` of your local computer.

Like all ``verdi`` commands, you can select a different AiiDA profile via the ``-p PROFILE`` option:

.. code-block:: bash

    verdi -p <another_profile> restapi

.. note::

    REST API version history:

     * ``aiida-core`` >= 1.0.0b6: ``v4``
     * ``aiida-core`` >= 1.0.0b3, <1.0.0b6: ``v3``
     * ``aiida-core`` <1.0.0b3: ``v2``

> 

Reviewer: Please add a link here:
```
.. _how-to:data:serve:query:
```
I'd like to reference it

Author: done!

The default traversal rules make sure to include the complete provenance of any node specified and should be sufficient for most cases.
See ``verdi export create -h`` for ways to modify the traversal rules.

.. tip::

    If you want to make sure that the archive includes everything you intended, you can create a new profile and import it:

    .. code-block:: console

            verdi quicksetup --profile test-export  # create new profile
            verdi --profile test-export import my-calculations.aiida

            # now use e.g. the QueryBuilder to query the database

Please remember to use **UUIDs** when pointing your colleagues to data *inside* an AiiDA export file, since UUIDs are guaranteed to be universally unique (while PKs aren't).

Using groups for data exporting
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If the number of results to be exported is large, for example in a high-throughput study, use the ``QueryBuilder`` to add the corresponding nodes to a group ``my-results`` (see :ref:`how-to:data:organize:group`).
Then export the group:

.. code-block:: console

    $ verdi export create my-calculations.aiida --groups my-results

> Publishing AiiDA export files

Reviewer 2: ```suggestion
Publishing AiiDA archive files
```

^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> AiiDA export files can be published on any research data repository, for example the `Materials Cloud Archive`_, `Zenodo`_, or the `Open Science Framework`_.

Reviewer 2: ```suggestion
AiiDA archive files can be published on any research data repository, for example the `Materials Cloud Archive`_, `Zenodo`_, or the `Open Science Framework`_.
```

> When publishing AiiDA export files on the `Materials Cloud Archive`_, you also get an interactive *EXPLORE* section, which allows peers to browse the AiiDA provenance graph directly in the browser.

Reviewer 2: Please find-and-replace all of these.
```suggestion
When publishing AiiDA archive files on the `Materials Cloud Archive`_, you also get an interactive *EXPLORE* section, which allows peers to browse the AiiDA provenance graph directly in the browser.
```

.. _Zenodo: https://zenodo.org
.. _Open Science Framework: https://osf.io
.. _Materials Cloud Archive: https://archive.materialscloud.org

Importing an archive
^^^^^^^^^^^^^^^^^^^^

Use ``verdi import`` to import AiiDA archives into your current AiiDA profile.
``verdi import`` accepts URLs, e.g.:

.. code-block:: console

    $ verdi import "https://archive.materialscloud.org/record/file?file_id=2a59c9e7-9752-47a8-8f0e-79bcdb06842c&filename=SSSP_1.1_PBE_efficiency.aiida&record_id=23"

> During import, AiiDA will avoid identifier collisions and node duplication based on UUIDs (and email comparisons for ``User`` entries).

Reviewer 2: Maybe not necessary, since the entries will not have anything directly to do with the Python representation of the same.
```suggestion
During import, AiiDA will avoid identifier collisions and node duplication based on UUIDs (and email comparisons for :py:class:~`aiida.orm.users.User` entries).
```

By default, existing entities will be updated with the most recent changes.

> Node extras and comments have special modes for determining how to import them - for more details, see ``verdi import -h``.

Reviewer 2: ```suggestion
Node extras and comments have special modes for determining how to import them - for more details, see ``verdi import --help``.
```

.. tip:: The AiiDA archive format has evolved over time, but you can still import archives created with previous AiiDA versions.

>     If an outdated archive version is detected during import, you will be prompted to confirm automatic migration of the archive file.

Reviewer 2: Does this work with multi-line + in-line?

Author: thanks, I checked and it works fine.

    This feature is, however, part of the `AiiDA roadmap <https://github.com/aiidateam/aiida-core/wiki/AiiDA-release-roadmap>`_.

Launching the REST API
^^^^^^^^^^^^^^^^^^^^^^

Start serving data from your default AiiDA profile via the REST API:

.. code-block:: console

    $ verdi restapi
     * REST API running on http://127.0.0.1:5000/api/v4
     * Serving Flask app "aiida.restapi.run_api" (lazy loading)
     * Environment: production
       WARNING: This is a development server. Do not use it in a production deployment.
       Use a production WSGI server instead.
     * Debug mode: off
     * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)

The REST API is now running on port ``5000`` of your local computer.

Like all ``verdi`` commands, you can select a different AiiDA profile via the ``-p PROFILE`` option:

.. code-block:: bash

    verdi -p <another_profile> restapi

.. note::

    REST API version history:

     * ``aiida-core`` >= 1.0.0b6: ``v4``
     * ``aiida-core`` >= 1.0.0b3, <1.0.0b6: ``v3``
     * ``aiida-core`` <1.0.0b3: ``v2``

Querying the REST API
^^^^^^^^^^^^^^^^^^^^^

> 

Reviewer 2: ```suggestion
```

> Open the base URL of the REST API (by default: `http://127.0.0.1:5000/api/v4 <https://aiida-dev.materialscloud.org/2dstructures/api/v4/nodes>`_) in your web browser in order to get an overview of the available endpoints.

Reviewer 2: Why mislead here and point to a MaterialsCloud REST API if you're showing the URL of something different?
Perhaps make it more clear that you're sending people to MaterialsCloud
```suggestion
Open the base URL of the REST API (by default: http://127.0.0.1:5000/api/v4, `example on MaterialsCloud <https://aiida-dev.materialscloud.org/2dstructures/api/v4/nodes>`_) in your web browser in order to get an overview of the available endpoints.
```

Author: This was a copy/paste error, will fix this

For an extensive user documentation, see the :ref:`AiiDA REST API reference <reference:rest-api>`.

> 

Reviewer 2: ```suggestion
```

.. tip::

    The interactive `EXPLORE sections on Materials Cloud <https://www.materialscloud.org/explore/menu>`_ are all powered by the AiiDA REST API and you can query the underlying API, either using your web browser or using a tool like ``curl``:

    .. code-block:: console

       $ curl https://aiida-dev.materialscloud.org/2dstructures/api/v4/users

> 

Reviewer 2: ```suggestion
```

.. note::
    One Apache/NGINX server can host multiple APIs, e.g. connecting to different AiiDA profiles.

In the following, we assume you have a working installation of Apache with the ``mod_wsgi`` `WSGI module <modwsgi.readthedocs.io/>`_ enabled.

The goal of the example is to hookup the APIs ``django`` and ``sqlalchemy`` pointing to two AiiDA profiles, called for simplicity ``django`` and ``sqlalchemy``.

All the relevant files are enclosed under the path ``/docs/wsgi/`` starting from the AiiDA source code path.
In each of the folders ``app1/`` and ``app2/``, there is a file named ``rest.wsgi`` containing a python script that instantiates and configures a python web app called ``application``, according to the rules of ``mod_wsgi``.
For how the script is written, the object ``application`` is configured through the file ``config.py`` contained in the same folder.
Indeed, in ``app1/config.py`` the variable ``aiida-profile`` is set to ``"django"``, whereas in ``app2/config.py`` its value is ``"sqlalchemy"``.

> Anyway, the path where you put the ``.wsgi`` file as well as its name are irrelevant as long as they are correctly referred to in the Apache configuration file, as shown later on.

Reviewer 2: ```suggestion
The path where you put the ``.wsgi`` file as well as its name are irrelevant as long as they are correctly referred to in the Apache configuration file, as shown later on.
```

Similarly, you can place ``config.py`` in a custom path, provided you change the variable ``config_file_path`` in the ``wsgi file`` accordingly.

> In ``rest.wsgi`` probably the only options you might want to change is ``catch_internal_server``.

Reviewer 2: ```suggestion
In ``rest.wsgi`` the only options you might want to change is ``catch_internal_server``.
```

When set to ``True``, it lets the exceptions thrown during the execution of the app propagate all the way through until they reach the logger of Apache.
Especially when the app is not entirely stable yet, one would like to read the full python error traceback in the Apache error log.

Finally, you need to setup the Apache site through a proper configuration file.
We provide two template files: ``one.conf`` or ``many.conf``.
The first file tells Apache to bundle both apps in a unique Apache daemon process.

> Apache usually creates multiple process dynamically and with this configuration each process will handle both apps.

Reviewer 2: ```suggestion
Apache usually creates multiple processes dynamically and with this configuration each process will handle both apps.
```

The script ``many.conf``, instead, defines two different process groups, one for each app.
So the processes created dynamically by Apache will always be handling one app each.
The minimal number of Apache daemon processes equals the number of apps, contrarily to the first architecture, where one process is enough to handle two or even a larger number of apps.

> Let us call the two apps for this example ``django`` and ``sqlalchemy``.

Reviewer 2: ```suggestion
Let us call the two apps for this example ``django`` and ``sqlalchemy``, matching with the chosen AiiDA profiles.
```

In both ``one.conf`` and ``many.conf``, the important directives that should be updated if one changes the paths or names of the apps are:

    - ``WSGIProcessGroup`` to define the process groups for later reference.
      In ``one.conf`` this directive appears only once to define the generic group ``profiles``, as there is only one kind of process handling both apps.
      In ``many.conf`` this directive appears once per app and is embedded into a "Location" tag, e.g.::

        <Location /django>
            WSGIProcessGroup sqlalchemy
        <Location/>

    - ``WSGIDaemonProcess`` to define the path to the AiiDA virtual environment.
      This appears once per app in both configurations.

    - ``WSGIScriptAlias`` to define the absolute path of the ``.wsgi`` file of each app.

    - The ``<Directory>`` tag mainly used to grant Apache access to the files used by each app, e.g.::

        <Directory "<aiida.source.code.path>/aiida/restapi/wsgi/app1">

>                 Require all granted

Reviewer 2: ```suggestion
            Require all granted
```

        </Directory>

The latest step is to move either ``one.conf`` or ``many.conf`` into the Apache configuration folder and restart the Apache server.
In Ubuntu, this is usually done with the commands:

.. code-block:: bash

    cp <conf_file>.conf /etc/apache2/sites-enabled/000-default.conf
    sudo service apache2 restart

We believe the two basic architectures we have just explained can be successfully applied in many different deployment scenarios.

> Nevertheless, we suggest users who need finer tuning of the deployment setup to look into to the official documentation of `Apache <https://httpd.apache.org/>`_ and, more importantly, `WSGI <wsgi.readthedocs.io/>`__.

Reviewer 2: ```suggestion
Nevertheless, we suggest users who need finer tuning of the deployment setup to look into to the official documentation of `Apache <https://httpd.apache.org/>`_ and, more importantly, `WSGI <wsgi.readthedocs.io/>`_.
```

The URLs of the requests handled by Apache must start with one of the paths specified in the directives ``WSGIScriptAlias``.
These paths identify uniquely each app and allow Apache to route the requests to their correct apps.
Examples of well-formed URLs are:

.. code-block:: bash

    curl http://localhost/django/api/v4/computers -X GET
    curl http://localhost/sqlalchemy/api/v4/computers -X GET

> The first (second) request will be handled by the app ``django`` (``sqlalchemy``), namely will serve results fetched from the profile ``django`` (``sqlalchemy``).

Reviewer 2: ```suggestion
The first (second) request will be handled by the app ``django`` (``sqlalchemy``), and will serve results fetched from the AiiDA profile ``django`` (``sqlalchemy``).
```

> Notice that we haven't specified any port in the URLs since Apache listens conventionally to port 80, where any request lacking the port is automatically redirected.

Reviewer 2: ```suggestion
Notice that we have not specified any port in the URLs since Apache listens conventionally to port 80, where any request lacking the port is automatically redirected (port 443 for HTTPS).
```

.. _how-to:share:

*****************
How to share data
*****************

AiiDA offers two avenues for sharing data with others: archive files and the REST API.

.. _how-to:share:archives:

Sharing AiiDA archives
======================

You have performed your calculations with AiiDA and you would like to share your AiiDA provenance graph, for example to make your scientific study reproducible.

Since AiiDA keeps track of the provenance of every computed result, this step is easy:

> tell AiiDA the **final results** you would like to be reproducible, and AiiDA will automatically include their entire provenance using the :ref:`topics:provenance:consistency:traversal-rules`.

Reviewer 2: ```suggestion
Tell AiiDA the **final results** you would like to be reproducible, and AiiDA will automatically include their entire provenance using the :ref:`topics:provenance:consistency:traversal-rules`.
```

Exporting individual nodes
^^^^^^^^^^^^^^^^^^^^^^^^^^
Let's say the key results of your study are contained in three AiiDA nodes with PKs ``12``, ``123``, ``1234``.
Exporting those results together with their provenance is as easy as:

.. code-block:: console

    $ verdi export create my-calculations.aiida --nodes 12 123 1234

As usual, you can use any identifier (label, PK or UUID) to specify the nodes to be exported.

The resulting archive file ``my-calculations.aiida`` contains all information pertaining to the exported nodes.
The default traversal rules make sure to include the complete provenance of any node specified and should be sufficient for most cases.

> See ``verdi export create -h`` for ways to modify the traversal rules.

Reviewer 2: ```suggestion
See ``verdi export create --help`` for ways to modify the traversal rules.
```

.. tip::

>     If you want to make sure that the archive includes everything you intended, you can create a new profile and import it:

Reviewer 2: ```suggestion
    If you want to make sure the archive includes everything you intended, you can create a new profile and import it:
```

    .. code-block:: console

>             verdi quicksetup --profile test-export  # create new profile
>             verdi --profile test-export import my-calculations.aiida
> 
>             # now use e.g. the QueryBuilder to query the database

Reviewer 2: ```suggestion
        $ verdi quicksetup --profile test-export  # create new profile
        $ verdi --profile test-export import my-calculations.aiida

    Now use, e.g. the :py:class:`~aiida.orm.querybuilder.QueryBuilder` to query the database.
```

> Please remember to use **UUIDs** when pointing your colleagues to data *inside* an AiiDA export file, since UUIDs are guaranteed to be universally unique (while PKs aren't).

Reviewer 2: Again, please perform a search-and-replace for these cases.
```suggestion
Please remember to use **UUIDs** when pointing your colleagues to data *inside* an AiiDA archive file, since UUIDs are guaranteed to be universally unique (while PKs aren't).
```

Reviewer 2: Did you perform a search-and-replace for these cases? Or are you sure all cases have been addressed?

Author: Ok, I'll do a search & replace on the AiiDA codebase

Reviewer 2: I was thinking more within this document (and the documentation in general), but uhm, sure 😅 But maybe this can also wait for a separate PR?

Reviewer: I think we can do this once we add the new implementation. We will anyway have to rewrite parts of the documentation. Wouldn't hold up this PR for that

Reviewer 2: Indeed - but it seems @ltalirz is too fast for us both - already went and did it 😮

Using groups for data exporting
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

If the number of results to be exported is large, for example in a high-throughput study, use the ``QueryBuilder`` to add the corresponding nodes to a group ``my-results`` (see :ref:`how-to:data:organize:group`).
Then export the group:

.. code-block:: console

    $ verdi export create my-calculations.aiida --groups my-results

Publishing AiiDA archive files
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

AiiDA archive files can be published on any research data repository, for example the `Materials Cloud Archive`_, `Zenodo`_, or the `Open Science Framework`_.
When publishing AiiDA archive files on the `Materials Cloud Archive`_, you also get an interactive *EXPLORE* section, which allows peers to browse the AiiDA provenance graph directly in the browser.

> 

Reviewer 2: ```suggestion
```

.. _Zenodo: https://zenodo.org
.. _Open Science Framework: https://osf.io
.. _Materials Cloud Archive: https://archive.materialscloud.org

Importing an archive
^^^^^^^^^^^^^^^^^^^^

Use ``verdi import`` to import AiiDA archives into your current AiiDA profile.
``verdi import`` accepts URLs, e.g.:

.. code-block:: console

    $ verdi import "https://archive.materialscloud.org/record/file?file_id=2a59c9e7-9752-47a8-8f0e-79bcdb06842c&filename=SSSP_1.1_PBE_efficiency.aiida&record_id=23"

During import, AiiDA will avoid identifier collisions and node duplication based on UUIDs (and email comparisons for :py:class:~`aiida.orm.users.User` entries).
By default, existing entities will be updated with the most recent changes.
Node extras and comments have special modes for determining how to import them - for more details, see ``verdi import --help``.

.. tip:: The AiiDA archive format has evolved over time, but you can still import archives created with previous AiiDA versions.
    If an outdated archive version is detected during import, you will be prompted to confirm automatic migration of the archive file.

>     You can also use ``verdi export migrate`` to update existing archive files once and for all.

Reviewer 2: ```suggestion
    You can also use ``verdi export migrate`` to create updated archive files from existing archive files (or update them in place).
```

.. _how-to:share:serve:

Serving data through the REST API
=================================

The AiiDA REST API allows to query your AiiDA database over HTTP(S).

.. note::

    As of October 2020, the AiiDA REST API does *not* yet support workflow management.
    This feature is, however, part of the `AiiDA roadmap <https://github.com/aiidateam/aiida-core/wiki/AiiDA-release-roadmap>`_.

.. _how-to:share:serve:launch:

Launching the REST API
^^^^^^^^^^^^^^^^^^^^^^

Start serving data from your default AiiDA profile via the REST API:

.. code-block:: console

    $ verdi restapi
     * REST API running on http://127.0.0.1:5000/api/v4
     * Serving Flask app "aiida.restapi.run_api" (lazy loading)
     * Environment: production
       WARNING: This is a development server. Do not use it in a production deployment.
       Use a production WSGI server instead.
     * Debug mode: off
     * Running on http://127.0.0.1:5000/ (Press CTRL+C to quit)

The REST API is now running on port ``5000`` of your local computer.

Like all ``verdi`` commands, you can select a different AiiDA profile via the ``-p PROFILE`` option:

.. code-block:: bash

    verdi -p <another_profile> restapi

> 

Reviewer 2: ```suggestion
```
